constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java

```diff
@@ -0,0 +1,359 @@
+            setAngles(mReferenceAngles);
+        }
+        if (mReferenceRadius != null) {
+            setRadius(mReferenceRadius);
+        }
+        anchorReferences();
+    }
+
+    private void anchorReferences() {
+        mContainer = (ConstraintLayout) getParent();
+        ConstraintSet c = new ConstraintSet();
+        c.clone(mContainer);
+        for (int i = 0; i <= mCount; i++) {
+            int id = mIds[i];
+            View view = mContainer.getViewById(id);
+
+            if (view != null) {
+                int radius = DEFAULT_RADIUS;
+                float angle = DEFAULT_ANGLE;
+
+                if (i < getRadius().length){
+                    radius = getRadius()[i];
+                } else {
+                    Log.e("CircularFlow", "Added radius to view with id: " + mMap.get(view.getId()));
+                }
+
+                if (i < getAngles().length){
+                    angle = getAngles()[i];
+                } else {
+                    Log.e("CircularFlow", "Added angle to view with id: " + mMap.get(view.getId()));
+                }
+                c.constrainCircle(view.getId(), mViewCenter, radius, angle);
+            }
+        }
+        c.applyTo(mContainer);
+        applyLayoutFeatures();
+    }
+
+    /**
+     * Add a view to the CircularFlow. The referenced view need to be a child of the container parent.
+     * The view also need to have its id set in order to be added.
+     * The views previous need to have its radius and angle set in order to be added correctly a new view.
+     * @param view
+     * @param radius
+     * @param angle
+     * @return
+     */
+    public void addViewToCircularFlow(View view, int radius, float angle) {
+        if (containsId(view.getId())){
+            return;
+        }
+        addView(view);
+        mCountAngle++;
+        mAngles = getAngles();
+        mAngles[mCountAngle - 1] = angle;
+        mCountRadius++;
+        mRadius = getRadius();
+        mRadius[mCountRadius - 1] = (int) (radius * myContext.getResources().getDisplayMetrics().density);
+        anchorReferences();
+    }
+
+    /**
+     * Remove a given view from the CircularFlow.
+     *
+     * @param view
+     */
+    public void removeViewFromCircularFlow(View view){
+        int index = removeView(view);
+        if (index == -1) {
+            return;
+        }
+        ConstraintSet c = new ConstraintSet();
+        c.clone(mContainer);
+        c.clear(view.getId(), ConstraintSet.CIRCLE_REFERENCE);
+        c.applyTo(mContainer);
+
+        if (index < mAngles.length) {
+            mAngles = removeAngle(mAngles, index);
+            mCountAngle--;
+        }
+        if (index < mRadius.length) {
+            mRadius = removeRadius(mRadius, index);
+            mCountRadius--;
+        }
+        anchorReferences();
+    }
+
+
+    /**
+     * @hide
+     */
+    private float[] removeAngle(float[] angles, int index) {
+        if (angles == null
```

Comment:
small suggestion: refactor in a utility function to remove an element of a given array and call it from removeAngle() and removeRadius()

Reply:
Refactor done ✅